### PR TITLE
v10: ensure blocklist supports dynamic label expressions

### LINKED ISSUE #10127: Block list very slow to / fails to render labels when using  {{ propertyName | ncNodeName}}
Block list very slow to / fails to render labels when using  `{{ propertyName | ncNodeName}}`

## Umbraco version

I am seeing this issue on Umbraco version: 8.12.2

Reproduction
------------

### Bug summary

When using the block list with a block that contains a multinode tree picker, I want to display the name of the selected node on the block, similar to if I had used the multinode treepicker directly and allowed multiple items. The node name on each block either loads very slowly or does not load at all.

image

### Specifics

Google Chrome on Windows - latest

### Steps to reproduce

Create a document type with a block list editor property.
The block list editor contains one block with a multinode treepicker on the content model with alias = article.
The block has the label expression
`{{ article | ncNodeName }}`

Create / publish the above then add multiple items to the block list editor.

### Expected result

Labels should reliably and quickly update to show the name of the article selected on the content model.

### Actual result

Label fails to load or loads on the order of minutes the name of the article selected on the content model.

---
_This item has been added to our backlog [AB#21097](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/21097)_

### PR DESCRIPTION
**Pull request by iOvergaard**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

Fixes #10127 

### Description

We have the ability to set a custom label expression for blocks in the block list editor. This expression can even point to the element type itself and extract values from whatever is in the block, e.g. `{{ description }}` if the element type has a "description" property.

Sometimes, you might want to extract more data from your properties. Say you have an MNTP with articles and you wanted to display the name of that article. This was possible in the Nested Content editor, but it has had its ups and downs in the block list. Nominally you do this by adding the `ncNodeName` filter to your expression like this: `{{ article | ncNodeName }}`. The thing is that this filter requires an async operation in most cases unless you have the article picked several places.

The first change proposed in this PR is to mark that filter as `$stateful` and add some checks on fallthrough cases where the same node is requested twice in a row before the async operation completes, null-ish node IDs, etc.

The next change is to remove the `$interpolate` logic for custom labels and make sure that the template for blocks always compiles the label instead, which in turn ensures that stateful filters are being updated ($interpolate seems to ignore filters that are stateful, even though other direct values are updated just fine).

To make the second change work, this PR introduces a new directive called `blockHtmlCompile` which you can input any custom expression and it will compile that against the former interpolatable values of each block, including some values that are set dynamically such as $index. We can then use it like this for example in labelblock.editor.html:

```html
<span block-html-compile="block.label" block-html-scope="block"></span>
```

[See this Plunker](https://plnkr.co/edit/Bc3xzETZChRLE9Sj) for what works and what does not.

If desired this new compile directive could be abstracted into a generic one and used elsewhere in the future.

Everything put together works like this:

block-ncnodename

### Steps to reproduce
Create a document type with a block list editor property.
The block list editor contains one block with a multinode treepicker on the content model with alias = article.
The block has the label expression `{{ article | ncNodeName }}`

Create / publish the above then add multiple items to the block list editor.

---
_This item has been added to our backlog [AB#21874](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/21874)_

### CODE REVIEW

File: src/Umbraco.Web.UI.Client/src/common/filters/nestedcontent.filter.js [hunk 1]
```diff
@@ -3,67 +3,78 @@
...
+    if (serviceInvoked) {
+      return formatLabel(ncNodeNameCache.keys[lookupId], ids.length);
+    }
 
-        // See if there is a value in the cache and use that
-        if (ncNodeNameCache.keys[lookupId]) {
-            return formatLabel(ncNodeNameCache.keys[lookupId], ids.length);
+    serviceInvoked = true;
+
+    var type = lookupId.indexOf("umb://media/") === 0
```
Comment (bjarnef) [inline src/Umbraco.Web.UI.Client/src/common/filters/nestedcontent.filter.js:58]: @iOvergaard should we use `udiParser.parse()` to get entity type?
https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Web.UI.Client/src/common/services/udiParser.service.js#L24

Comment (iOvergaard) [reply src/Umbraco.Web.UI.Client/src/common/filters/nestedcontent.filter.js:58]: You are right, @bjarnef. We should definitely do that. I guess there are some leftovers in the code base that don't follow the service pattern. I will update this instance even though it doesn't directly relate to this PR.

Comment (iOvergaard) [reply src/Umbraco.Web.UI.Client/src/common/filters/nestedcontent.filter.js]: Done

**Comment by iOvergaard** [comment]
I have created a Plunker to showcase the concept: https://plnkr.co/edit/Bc3xzETZChRLE9Sj

**Comment by nielslyngsoe** [comment]
Thanks @iOvergaard for discovering the option of using $compile to render the expression, this is defiantly the right way to enable filters to be stateful.

As I reviewed this PR it turned out that several things were broken by this approach and it was a breaking-change as custom views using block.label would be broken.

Main things that didn't work:
- Block List Editor within Block List Editor would not synchronize data in time to be persisted.
- Block.label was not an option anymore, which as well broke the clipboard items label.

The base of this PR is great, but as I ended up transforming most of the parts I have created a new PR with the changes I made. (To be reviewed by someone else) Therefore I'm closing this one.

https://github.com/umbraco/Umbraco-CMS/pull/12909

**Pull request closed without merging**